Improve deprecation message for issue 8368

sbidoul opened the pull request:
When `setup.py install` fails and `setup.py bdist_wheel` failed too, it is most probably for the same reason.
Reporting a deprecation warning about the failed wheel build in that case is causing noise in reports for #8368. What we are really interested in is reports about packages that can install via `setup.py install` but fail building a wheel.

This PR makes the deprecation look like this:

```console
$ pip install -f pip/tests/data/packages/ wheelbroken
Looking in links: pip/tests/data/packages/
Processing ./pip/tests/data/packages/wheelbroken-0.1.tar.gz
Building wheels for collected packages: wheelbroken
  Building wheel for wheelbroken (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/sbi-local/.virtualenvs/tempenv-04a2234147f66/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-7y0wc2c_/wheelbroken/setup.py'"'"'; __file__='"'"'/tmp/pip-install-7y0wc2c_/wheelbroken/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-gswzfp2o
       cwd: /tmp/pip-install-7y0wc2c_/wheelbroken/
  Complete output (5 lines):
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-install-7y0wc2c_/wheelbroken/setup.py", line 8, in <module>
      raise FakeError('this package designed to fail on bdist_wheel')
  __main__.FakeError: this package designed to fail on bdist_wheel
  ----------------------------------------
  ERROR: Failed building wheel for wheelbroken
  Running setup.py clean for wheelbroken
Failed to build wheelbroken
Installing collected packages: wheelbroken
    Running setup.py install for wheelbroken ... done
  DEPRECATION: wheelbroken was installed using the legacy 'setup.py install' method, because a wheel could not be built for it. pip 21.0 will remove support for this functionality. A possible replacement is to fix the wheel build issue reported above. You can find discussion regarding this at https://github.com/pypa/pip/issues/8368.
Successfully installed wheelbroken-0.1
```

uranusjr commented:
Do we rollback previous `setup.py install` changes when a later package fails to install? (I believe wheel installations are rolled back.) Of that’s the case, maybe it’d be a better idea to show only one message when all installations succeed, instead of one message for each package?

sbidoul commented:
> Do we rollback previous setup.py install changes when a later package fails to install

@uranusjr I don't think so. I tried `pip install -f pip/tests/data/packages/ wheelbroken psutil`  where `setup.py install` of wheelbroken succeeds and `setup.py install` of psutil fails and I end up with only `wheelbroken` installed and the deprecation warning for it only.

uranusjr commented:
That’s unfortunately, although I’m not surprised since `setup.py install` is difficult to manage (which is why we want it to go away in the first place). I guess this is the best approach then.

uranusjr commented:
@pradyunsg Would it be possible to have one last 20.2 release for this, if nothing else comes up? I really don’t want to deal with the bogus issues for three months.

pradyunsg commented:
Sure. 👍🏽 

Someone needs to add a NEWS file here tho. :)

sbidoul commented:
I added the news file.

I also changed install success detection to consider the `success` flag instead of the absence of exception.

I must say I dont quite understand this block... :man_shrugging: 

https://github.com/pypa/pip/blob/ea10d0e253f497c9fc08097f14cdba6b94b45541/src/pip/_internal/req/req_install.py#L859-L861

pfmoore commented:
See here: https://github.com/pypa/pip/blob/ea10d0e253f497c9fc08097f14cdba6b94b45541/src/pip/_internal/req/__init__.py#L92-L96

If the install didn't succeed, we roll back, which we can't do for a legacy install... It's a nasty bodge to use the success flag like that, but that's true of a lot of pip's code 🙂

sbidoul commented:
Thanks for the explanation @pfmoore :see_no_evil: :hear_no_evil: :speak_no_evil: :)

It's :green_circle: now.

pradyunsg commented:
I'm gonna merge and release this sometime tomorrow, if no one has holler'd here to ask me to not do it. :)

pradyunsg commented:
I'd set the calendar reminder for the wrong week. :)

Muheeb42318 commented:
I get an error while installing librec. Can somebody help? It says "Error while installing librec- "failed building wheel for librec "

ERROR: Failed building wheel for librec
Running setup.py clean for librec
Failed to build librec
DEPRECATION: Could not build wheels for librec which do not use PEP 517. pip wil
l fall back to legacy 'setup.py install' for these. pip 21.0 will remove support
for this functionality. A possible replacement is to fix the wheel build issue
reported above. You can find discussion regarding this at https://github.com/pyp
a/pip/issues/8368.
Installing collected packages: librec

error: command 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Buil
dTools\VC\Tools\MSVC\14.27.29110\bin\HostX86\x64\cl.exe' failed with exi
t status 2
